tools/generator/Utilities/TypeNameUtilities.cs

```diff
@@ -101,5 +102,41 @@ public static string StudlyCase (string name)
 			}
 			return builder.ToString ();
 		}
+
+		// From https://github.com/xamarin/xamarin-android/blob/master/src/Xamarin.Android.Build.Tasks/Utilities/ResourceIdentifier.cs
+		// https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#identifiers
+		private const string FormattingCharacter = @"\p{Cf}";
+		private const string ConnectingCharacter = @"\p{Pc}";
+		private const string DecimalDigitCharacter = @"\p{Nd}";
+		private const string CombiningCharacter = @"\p{Mn}\p{Mc}";
+		private const string LetterCharacter = @"\p{Lu}\p{Ll}\p{Lt}\p{Lm}\p{Lo}\p{Nl}";
+
+		private const string IdentifierPartCharacter = LetterCharacter +
+		    DecimalDigitCharacter +
+		    ConnectingCharacter +
+		    CombiningCharacter +
+		    FormattingCharacter;
+
+		private const string IdentifierStartCharacter = "(" + LetterCharacter + "_)";
+
+		private const string Identifier = IdentifierStartCharacter + "(" + IdentifierPartCharacter + ")";
+
+		// We use [^ ...] to detect any character that is NOT a match.
+		static Regex validIdentifier = new Regex ($"[^{Identifier}]", RegexOptions.Compiled);
+
+		public static string CreateValidIdentifier (string identifier, bool useEncodedReplacements = false)
```

Comment:
Could we move this into `Java.Interop.Tools.JavaCallableWrappers` or some other library referenced by `generator.csproj` and also referenced by `Xamarin.Android.Build.Tasks.dll`?  That way a future xamarin-android PR could unify onto this method, instead of us having two "separate" methods which do the same thing.

Having the code in `generator.exe` slightly complicates that.